### PR TITLE
Expose mixed_precision dtype arguments

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #345
* #344
* #318
* __->__ #348

add training.mixed_precision_param and .mixed_precision_reduce options

refactor a util to map strings to torch dtypes

### CODE REVIEW

File: torchtitan/parallelisms/parallelize_llama.py [hunk 1]
```diff
@@ -209,9 +209,9 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
     if parallel_dims.dp_enabled:
         dp_mesh = world_mesh["dp"] if world_mesh.ndim > 1 else world_mesh
         assert dp_mesh.mesh_dim_names == ("dp",), dp_mesh.mesh_dim_names
-        # TODO: Expose `reduce_dtype` as a config option.
         mp_policy = MixedPrecisionPolicy(
-            param_dtype=torch.bfloat16, reduce_dtype=torch.float32
+            param_dtype=job_config.training.mixed_precision_param,
+            reduce_dtype=job_config.training.mixed_precision_param,
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py]:
```suggestion
            reduce_dtype=job_config.training.mixed_precision_reduce,
```

File: torchtitan/config_manager.py [hunk 2]
```diff
@@ -9,13 +9,31 @@
...
+DTYPE_MAP = {
+    "float16": torch.float16,
+    "float32": torch.float32,
+    "bfloat16": torch.bfloat16,
+}
+
+TORCH_DTYPE_ARGS = [
+    "checkpoint.export_dtype",
+    "training.mixed_precision_param",
+    "training.mixed_precision_reduce",
```
Comment (kwen2501) [inline torchtitan/config_manager.py]: nit: should "reduce" be "grad"?

File: torchtitan/config_manager.py [hunk 3]
```diff
@@ -207,6 +225,26 @@ def __init__(self):
...
+            """,
+        )
+        self.parser.add_argument(
+            "--training.mixed_precision_reduce",
+            type=torch_dtype,
+            default=torch_dtype("float32"),
+            choices=["float32"],
+            help="""
+                torch dtype to use for reductions when applying mixed precision via FSDP.
+                This feature only takes effect when data_parallel_degree > 1
```
Comment (kwen2501) [inline torchtitan/config_manager.py:235]: nit: "reductions" -> "gradients"

File: torchtitan/config_manager.py [hunk 4]
```diff
@@ -361,6 +400,9 @@ def parse_args(self, args_list: list = sys.argv[1:]):
             try:
                 with open(config_file, "rb") as f:
                     for k, v in tomllib.load(f).items():
+                        for k_, v_ in v.items():
+                            if ".".join([k, k_]) in TORCH_DTYPE_ARGS:
+                                v[k_] = torch_dtype(v_)
```
Comment (kwen2501) [inline torchtitan/config_manager.py]: nit: comment please?

**Review by kwen2501 — APPROVED** [review]
Thanks for adding this.

File: torchtitan/config_manager.py (hunk 2 above)
Comment (wconstab) [reply torchtitan/config_manager.py]: no, im following the existing naming for the mixed_precision config struct

**Review by wanchaol — APPROVED** [review]
lgtm

**Pull request merged**